### PR TITLE
[FLINK-1688] Socket client sink added

### PR DESCRIPTION
**Pull request by miguel0afd**

### CODE REVIEW

File: flink-staging/flink-streaming/flink-streaming-connectors/src/main/java/org/apache/flink/streaming/connectors/socket/SocketClientSink.java [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+			if(LOG.isErrorEnabled()){
+				LOG.error("Cannot send message to socket server at {}:{}", hostName, port);
+			}
+		}
+	}
+
+	/**
+	 * Closes the connection.
+	 */
+	private void closeConnection(){
```
Comment (mxm) [inline flink-staging/flink-streaming/flink-streaming-connectors/src/main/java/org/apache/flink/streaming/connectors/socket/SocketClientSink.java]: really? :)

File: flink-staging/flink-streaming/flink-streaming-connectors/src/main/java/org/apache/flink/streaming/connectors/socket/SocketClientSink.java [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+	public SocketClientSink(String hostName, int port, SerializationSchema<IN, byte[]> schema) {
+		this.hostName = hostName;
+		this.port = port;
+		this.scheme = schema;
+	}
+
+	/**
+	 * Initializes the connection to Socket.
+	 */
+	public void initialize() {
```
Comment (mxm) [inline flink-staging/flink-streaming/flink-streaming-connectors/src/main/java/org/apache/flink/streaming/connectors/socket/SocketClientSink.java]: really? :) You could call this method `intializeConnection` like its counterpart `closeConnection`.

File: flink-staging/flink-streaming/flink-streaming-connectors/src/main/java/org/apache/flink/streaming/connectors/socket/SocketClientSink.java [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+import java.io.OutputStream;
+import java.net.Socket;
+
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.streaming.api.function.sink.RichSinkFunction;
+import org.apache.flink.streaming.connectors.util.SerializationSchema;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class SocketClientSink<IN> extends RichSinkFunction<IN> {
```
Comment (mxm) [inline flink-staging/flink-streaming/flink-streaming-connectors/src/main/java/org/apache/flink/streaming/connectors/socket/SocketClientSink.java]: Can you add a docstring for the class to explain its purpose?

**Comment by mxm** [comment]
Thanks for the pull request. Could you add some javadoc to the class and add an entry in the Flink website docs [1]?

[1] http://ci.apache.org/projects/flink/flink-docs-master/programming_guide.html#data-sinks

**Comment by miguel0afd** [comment]
Documentation patch sent: https://issues.apache.org/jira/browse/FLINK-1688

**Comment by miguel0afd** [comment]
New suggestions added to the code

**Comment by mbalassi** [comment]
Thanks, LGTM. Merging it tomorrow.

**Comment by hsaputra** [comment]
@mbalassi, @miguel0afd seemed like the PR contains more commits than the one targeted for the patch.

Be careful when merging it bc it could murky the history.

I would suggest create clean branch and do rebase instead of merge to keep history clean.

**Comment by mbalassi** [comment]
@hsaputra thanks for the heads up - I am cherry-picking and squashing the relevant commits.

**Comment by miguel0afd** [comment]
Sorry for the mess, I made a mistake with a undesired merge... Anyway, I can create a new branch in my fork repository and create a new pull request if that helps.

**Comment by mbalassi** [comment]
No problem, thanks for the contribution. Cherry-picked and squashed your commits. I've decided to make three changes:

1. The output was not necessarily flushed upon closure, fixed that.
2. Some lines where indented with spaces.
3. I moved the function into streaming-core - as the socket source is considered a core functionality I found this solution more consistent.

Merging as soon as travis comes in green.

**Pull request closed without merging**